### PR TITLE
Post to slack workflow

### PR DESCRIPTION
**Pull request by ghost**
Implementation:
  - Workflows:
    - invoke simple HTTP request
    - post messages to SLACK
  - Workers:
    - simple HTTP task, used in mentioned workflows
  - Services:
    - HTTP service, helper service used by worker

### CODE REVIEW

File: frinx/workflows/http_workflows/post_to_slack.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    name: str = 'Post_to_Slack'
+    version: int = 1
+    description: str = 'Post a message to your favorite Slack channel'
+    restartable: bool = True
+    labels: list[str] = ['SLACK', 'HTTP']
+    schema_version: int = 2
+    workflow_status_listener_enabled: bool = False
+
+    def workflow_builder(self, workflow_inputs: HTTPWorkflowService.GenericRequest.WorkflowInput) -> None:
+    #|-----------------------------------------------------------------------------------|
```
Comment (Jozefiel) [inline frinx/workflows/http_workflows/post_to_slack.py]: remove  it

File: frinx/workflows/http_workflows/post_to_slack.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
+from frinx.common.workflow.task import SimpleTask
+from frinx.common.workflow.task import SimpleTaskInputParameters
+from frinx.common.workflow.task import TaskType
+from frinx.workers.http_workers.http_workers import Http
+from frinx.workflows.http_workflows.generic import HTTPWorkflowService
+
+
+class PostToSlack(HTTPWorkflowService.GenericRequest):
+#||=====================================================================================||
```
Comment (Jozefiel) [inline frinx/workflows/http_workflows/post_to_slack.py]: remove  it

File: frinx/workers/http_workers/http_workers.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+from frinx.common.worker.task_def import TaskDefinition
+from frinx.common.worker.task_def import TaskInput
+from frinx.common.worker.task_def import TaskOutput
+from frinx.common.worker.task_result import TaskResult
+from frinx.common.worker.worker import WorkerImpl
+from frinx.services.http_service import http_service
+
+
+class Http(ServiceWorkersImpl):
+    ###############################################################################
```
Comment (Jozefiel) [inline frinx/workers/http_workers/http_workers.py]: remove this comment line

File: frinx/services/http_service/http_service.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    @classmethod
+    @validator('data')
+    def encode_decode(cls, data: dict[str, Any] | list[Any] | str | None) -> Optional[
+    dict[str, Any] | list[Any] | str | None]:
+        if data:
+            try:
+               json.dumps(json.loads(data)) if isinstance(data, str) else json.dumps(data)
+            except JSONDecodeError as err:
+                raise InvalidJSONError(err.args)
+        return data
```
Comment (Jozefiel) [inline frinx/services/http_service/http_service.py]: we can move that validation to the worker level

File: frinx/services/http_service/http_service.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+class HTTPOutput(BaseModel):
+
+    class Config:
+        min_anystr_length = 1
+
+    code: int
+    data: dict[str, Any]
+    errors: list[str]
+    logs: Optional[list[str]] | Optional[str] | None = None
+    url: Optional[str] | None = None
```
Comment (Jozefiel) [inline frinx/services/http_service/http_service.py]: same as input

File: frinx/common/conductor_enums.py [hunk 6]
```diff
@@ -36,3 +36,25 @@ class RetryLogic(str, Enum):
...
+
+
+class ContentType(str, Enum):
+    APPLICATION_JSON = 'application/json'
+    APPLICATION_XML = 'application/xml'
+    APPLICATION_X_WWW_FORM_URLENCODED = 'application/x-www-form-urlencoded'
+    MULTIPART_FROM_DATA = 'multipart/form-data'
+    TEXT_CSV = 'text/csv'
+    TEXT_PLAIN = 'text/plain'
+    TEXT_XML = 'text/xml'
```
Comment (Jozefiel) [inline frinx/common/conductor_enums.py]: these ENUMS are not directly related to the conductor and should not be located here

File: frinx/workflows/http_workflows/post_to_slack.py [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+from frinx.common.workflow.task import TaskType
+from frinx.common.workflow.workflow import FrontendWFInputFieldType
+from frinx.common.workflow.workflow import WorkflowImpl
+from frinx.common.workflow.workflow import WorkflowInputField
+from frinx.workers.http_workers.http_workers import HTTPWorkersService
+
+SLACK_WEBHOOK_ID_DEFAULT = 'T05ECRXU1B7/B05FA6EUAKA/8EPw07KLIxOUAhojfg2Fx7DH'
+MESSAGE_TEXT_DEFAULT = 'Hello Slack!'
+
+Worker = HTTPWorkersService.GenericHTTPWorker
```
Comment (Jozefiel) [inline frinx/workflows/http_workflows/post_to_slack.py]: use HTTPWorkersService.GenericHTTPWorker directly in code
this approach is less readable

File: frinx/workflows/http_workflows/post_to_slack.py [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+    def workflow_builder(self, workflow_inputs: WorkflowInput) -> None:
+        worker_input = SimpleTaskInputParameters(
+            uri=f'https://hooks.slack.com/services/{workflow_inputs.slack_webhook_id.wf_input}',
+            body={'text': workflow_inputs.message_text.wf_input},
+            method='POST'
+        )
+
+        self.tasks.append(
+            SimpleTask(
+                task_reference_name=Worker().task_def.name,
```
Comment (Jozefiel) [inline frinx/workflows/http_workflows/post_to_slack.py]: not needed to construct object
use: 
name=HTTPWorkersService.GenericHTTPWorker

File: frinx/workflows/http_workflows/post_to_slack.py [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+                input_parameters=worker_input,
```
Comment (Jozefiel) [inline frinx/workflows/http_workflows/post_to_slack.py]: use: 
SimpleTaskInputParameters()

that object provide ability to validate compatibility during the registration

File: frinx/workflows/http_workflows/generic.py [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+                'method': workflow_inputs.method.wf_input,
+                'headers': workflow_inputs.headers.wf_input,
+                'body': workflow_inputs.body.wf_input,
+            }
+
+            self.tasks.append(
+                SimpleTask(
+                    name=HTTPWorkersService.GenericHTTPWorker,
+                    task_reference_name='http_task',
+                    input_parameters=SimpleTaskInputParameters(http_request=http_request),
```
Comment (Jozefiel) [inline frinx/workflows/http_workflows/generic.py]: is that ok? why you sending http_request object to the worker?

File: frinx/workflows/http_workflows/generic.py [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+            headers: WorkflowInputField = WorkflowInputField(
+                name='headers',
+                frontend_default_value={},
+                description='Request headers',
+                type=FrontendWFInputFieldType.TEXTAREA,
+            )
+
+            body: WorkflowInputField = WorkflowInputField(
+                name='body',
+                frontend_default_value={},
```
Comment (Jozefiel) [inline frinx/workflows/http_workflows/generic.py]: is that correctly rendered in UI? I mean is not a bad idea, but not sure if that works correctly on frontend

File: frinx/workflows/http_workflows/generic.py [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+                frontend_default_value='GET',
+                description='Request method',
+                # TODO: options = enum HTTPMethod instead of strings
+                options=['GET', 'PUT', 'POST', 'DELETE', 'PATCH'],
+                type=FrontendWFInputFieldType.SELECT,
+            )
+
+            headers: WorkflowInputField = WorkflowInputField(
+                name='headers',
+                frontend_default_value={},
```
Comment (Jozefiel) [inline frinx/workflows/http_workflows/generic.py]: is that correctly rendered in UI? I mean is not a bad idea, but not sure if that works correctly on frontend

File: frinx/common/conductor_enums.py [hunk 13]
```diff
@@ -36,3 +36,25 @@ class RetryLogic(str, Enum):
     FIXED = 'FIXED'
     EXPONENTIAL_BACKOFF = 'EXPONENTIAL_BACKOFF'
     LINEAR_BACKOFF = 'LINEAR_BACKOFF'
+
+
+class HttpMethod(str, Enum):
+    # CONNECT = 'CONNECT'
+    # OPTIONS = 'OPTIONS'
+    # TRACE = 'TRACE'
```
Comment (Jozefiel) [inline frinx/common/conductor_enums.py]: why are these options commented?

File: frinx/services/http_service/http_service.py [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+            case ContentType.TEXT_CSV:
+                ...
+            case ContentType.TEXT_PLAIN:
+                ...
+            case ContentType.TEXT_XML:
+                if not is_xml_valid(http_input['body']):
+                    raise IncompatibleFormError()
+
+    if http_input.get('accept', None):
+        kwargs['headers']['Accept'] = http_input['accept']
```
Comment (Jozefiel) [inline frinx/services/http_service/http_service.py]: can you explain idea? why we need specific parameter for that header in http_input

File: frinx/common/conductor_enums.py (hunk 13 above)
Comment (Jozefiel) [reply frinx/common/conductor_enums.py]: and also. not explicitly conductor related

File: frinx/services/http_service/http_service.py [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+                ...
+            case ContentType.MULTIPART_FROM_DATA:
+                ...
+            case ContentType.TEXT_CSV:
+                ...
+            case ContentType.TEXT_PLAIN:
+                ...
+            case ContentType.TEXT_XML:
+                if not is_xml_valid(http_input['body']):
+                    raise IncompatibleFormError()
```
Comment (Jozefiel) [inline frinx/services/http_service/http_service.py]: not sure if we want to handle that complex logic.
i think, we discuss before that we don't want to handle that and definitely not in services part.

Comment (ghost) [reply frinx/services/http_service/http_service.py]: Yes, maybe woker input should validate it directly. Or shouldn't validate either.
May i move this code block into worker input validator or remove it?

File: frinx/services/http_service/http_service.py (hunk 14 above)
Comment (ghost) [reply frinx/services/http_service/http_service.py]: Yes, this should be auth, sorry. I'll fix it.

File: frinx/common/conductor_enums.py (hunk 13 above)
Comment (ghost) [reply frinx/common/conductor_enums.py]: I forget to uncomment it, sorry. I'll fix it.

File: frinx/workflows/http_workflows/generic.py (hunk 12 above)
Comment (ghost) [reply frinx/workflows/http_workflows/generic.py]: Good point, I'll inspect it in more detail. Thank you.

File: frinx/workflows/http_workflows/generic.py (hunk 11 above)
Comment (ghost) [reply frinx/workflows/http_workflows/generic.py]: Good point, I'll inspect it in more detail. Thank you.

File: frinx/workflows/http_workflows/generic.py (hunk 10 above)
Comment (ghost) [reply frinx/workflows/http_workflows/generic.py]: Yes, it's old solution. I'll make a changes.

File: frinx/workflows/http_workflows/post_to_slack.py (hunk 9 above)
Comment (ghost) [reply frinx/workflows/http_workflows/post_to_slack.py]: Yes, I used it.

File: frinx/workflows/http_workflows/post_to_slack.py (hunk 8 above)
Comment (ghost) [reply frinx/workflows/http_workflows/post_to_slack.py]: I need to construct object, name is instance attribute.

File: frinx/workflows/http_workflows/post_to_slack.py (hunk 7 above)
Comment (ghost) [reply frinx/workflows/http_workflows/post_to_slack.py]: Ok, I'll make a changes. Thank you.

File: frinx/common/conductor_enums.py (hunk 6 above)
Comment (ghost) [reply frinx/common/conductor_enums.py]: Ok, May I move it to the woker directory? E.g. frinx/workers/http_workers/enums.py

File: frinx/workflows/http_workflows/post_to_slack.py (hunk 8 above)
Comment (Jozefiel) [reply frinx/workflows/http_workflows/post_to_slack.py]: i guess, object is created in the simpleTask class

chesk how is that done in generic workflow

File: frinx/common/conductor_enums.py (hunk 6 above)
Comment (Jozefiel) [reply frinx/common/conductor_enums.py]: maybe to services/http_service/enums.py

File: frinx/workers/http_workers/http_workers.py [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+            response = http_service.http_task(worker_input)
+            logs = f'{worker_input.method} {worker_input.uri} {response.status_code} {response.reason}'
+            logs = str(worker_input)
+
+            return TaskResult(
+                logs=logs,
+                status=TaskResultStatus.COMPLETED if response.ok else TaskResultStatus.FAILED,
+                output=self.WorkerOutput(
+                    status_code=response.status_code,
+                    response=response.text,
```
Comment (Jozefiel) [inline frinx/workers/http_workers/http_workers.py]: Is text suitable in any situation? 
I mean if the response is not a text but JSON or any different type.

File: frinx/workers/http_workers/http_workers.py [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+from frinx.services.http_service.enums import HttpMethod
+from frinx.workers.http_workers.util import BasicAuth
+
+
+class HTTPWorkersService(ServiceWorkersImpl):
+
+    class GenericHTTPWorker(WorkerImpl):
+
+        class WorkerDefinition(TaskDefinition):
+            name: str = 'generic_HTTP_worker'
```
Comment (Jozefiel) [inline frinx/workers/http_workers/http_workers.py]: use values from previous task definition             
"name": "HTTP_task",
"description": '{"description": "Generic http task", "labels": ["BASICS","HTTP"]}',

File: frinx/workers/http_workers/http_workers.py (hunk 16 above)
Comment (ghost) [reply frinx/workers/http_workers/http_workers.py]: OK, I complete this with function witch can returns response also in JSON.

File: frinx/workers/http_workers/http_workers.py (hunk 17 above)
Comment (ghost) [reply frinx/workers/http_workers/http_workers.py]: OK, it's fixed.

**Review by Jozefiel — APPROVED** [review]
